app/frontend/utils/formio-helpers.ts

```diff
@@ -11,6 +11,39 @@ export const singleRequirementFormJson = (requirementJson: IFormIORequirement):
   }
 }
 
+export const singleRequirementSubmissionData = (submissionData: any, requirementKey: string) => {
+  const sectionKey = Object.keys(submissionData.data).find((sk) => submissionData.data[sk][requirementKey])
+  if (!sectionKey) return { data: { [requirementKey]: null } }
+
+  return { data: { [requirementKey]: submissionData.data[sectionKey][requirementKey] } }
+}
+
 export const isFieldSetKey = (key: string) => {
   return key.split("|").length === 4
 }
+
+export const compareSubmissionData = (beforeSubmissionData: ISubmissionData, afterSubmissionData: ISubmissionData) => {
+  const changedFields: string[] = []
+
+  // Helper function to compare two values
+  const isDifferent = (value1: any, value2: any): boolean => {
+    return JSON.stringify(value1) !== JSON.stringify(value2)
+  }
+
+  for (const sectionKey in afterSubmissionData.data) {
+    if (afterSubmissionData.data.hasOwnProperty(sectionKey)) {
+      const afterSection = afterSubmissionData.data[sectionKey]
+      const beforeSection = beforeSubmissionData.data[sectionKey] || {}
+
+      for (const fieldKey in afterSection) {
+        if (afterSection.hasOwnProperty(fieldKey)) {
+          if (isDifferent(afterSection[fieldKey], beforeSection[fieldKey])) {
+            changedFields.push(fieldKey)
+          }
+        }
+      }
```

Comment:
this has a lot of nested if's. It can be a bit more readable if using inverse logic to early return

```
for (const sectionKey in afterSubmissionData.data) {
  if (!afterSubmissionData.data.hasOwnProperty(sectionKey)) continue

  const afterSection = afterSubmissionData.data[sectionKey]
  const beforeSection = beforeSubmissionData.data[sectionKey] || {}

  for (const fieldKey in afterSection) {
    if (!afterSection.hasOwnProperty(fieldKey) || !isDifferent(afterSection[fieldKey], beforeSection[fieldKey]))
      continue

    changedFields.push(fieldKey)
  }
}
```